File: device-app/ios-object-c/MsgViewController.m
//
//  MsgViewController.m
//  sample
//
//  Created by 김정도 on 2018. 1. 22..
//  Copyright © 2018년 Rationalowl. All rights reserved.
//

#import "MsgViewController.h"

@interface MsgViewController ()

@end

@implementation MsgViewController

@synthesize msgListView;
@synthesize inputMessageField;
@synthesize tView, messages;

#pragma mark -
#pragma mark IBActions
// view controller life cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    inputMessageField.text = @"input data";
    messages = [[NSMutableArray alloc] init];
    
    self.tView.delegate = self;
    self.tView.dataSource = self;
    
    //set delegate
    MinervaManager* minMgr = [MinervaManager getInstance];
    [minMgr setMessageDelegate:self];
    
}

- (void)didReceiveMemoryWarning {
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}




#pragma mark -
#pragma mark IBActions
//send message

- (IBAction)sendUpstreamMsg:(id)sender {
    //NSString* svcId = @"d0a83353281e4b678774a0efa44fdd82";
    NSString* serverId = @"1fe45769e24348bfa501c32032958483";
    NSString* msg = inputMessageField.text;
    MinervaManager* mgr = [MinervaManager getInstance];
    [mgr sendUpstreamMsg:msg serverRegId:serverId];
    
    NSString* displayStr = [NSString stringWithFormat:@"=>%@  ", msg];
    //[self.messages insertObject:displayStr atIndex:0];
    [self.messages addObject:displayStr];
    [self.tView reloadData];
    NSIndexPath *topIndexPath = [NSIndexPath indexPathForRow:messages.count-1 inSection:0];
    [self.tView scrollToRowAtIndexPath:topIndexPath
                      atScrollPosition:UITableViewScrollPositionMiddle
                              animated:YES];
}



- (IBAction)sendP2PMsg:(id)sender {
    NSString* msg = inputMessageField.text;
    NSMutableArray* devices = [[NSMutableArray alloc] init];
    [devices addObject:@"cf12c6b3c46e4e318b6e3c77b0590b9d"];
    MinervaManager* mgr = [MinervaManager getInstance];
    [mgr sendP2PMsg:msg devices:devices];
    
    NSString* displayStr = [NSString stringWithFormat:@"=>(p2p)%@  ", msg];
    //[self.messages insertObject:displayStr atIndex:0];
    [self.messages addObject:displayStr];
    [self.tView reloadData];
    NSIndexPath *topIndexPath = [NSIndexPath indexPathForRow:messages.count-1 inSection:0];
    [self.tView scrollToRowAtIndexPath:topIndexPath
                      atScrollPosition:UITableViewScrollPositionMiddle
                              animated:YES];
}


#pragma mark -
#pragma mark message delegate

-(void) onDownstreamMsgRecieved: (int) msgSize msgList : (NSArray*) msgList alarmIdx : (int) alarmIdx {
    NSLog(@"onMsgRecieved msg size = %d", msgSize);
    NSDictionary* msg;
    
    NSString* serverRegId;
    long serverTime;
    NSString* msgData;
    
    NSDateFormatter* format = [[NSDateFormatter alloc] init];
    [format setDateFormat:@"yyyy/MM/dd HH:mm:ss"];
    
    for(int i = 0; i < msgSize; i++) {
        msg = msgList[i];
        // message sender(app server)
        serverRegId = msg[@"sender"];
        // message sent time
        serverTime = [msg[@"serverTime"] longValue];
        NSDate* date = [NSDate dateWithTimeIntervalSince1970:(serverTime /1000)];
        msgData = msg[@"data"];
        NSString* displayStr = [NSString stringWithFormat:@"%@  [sent time:%@]", msgData, [format stringFromDate:date]];
        //[self.messages insertObject:displayStr atIndex:0];
        [self.messages addObject:displayStr];
    }
    
    [self.tView reloadData];
    NSIndexPath *topIndexPath = [NSIndexPath indexPathForRow:messages.count-1 inSection:0];
    [self.tView scrollToRowAtIndexPath:topIndexPath
                      atScrollPosition:UITableViewScrollPositionMiddle
                              animated:YES];
}


-(void) onP2PMsgRecieved: (int) msgSize msgList : (NSArray*) msgList alarmIdx : (int) alarmIdx {
    NSLog(@"onP2PMsgRecieved msg size = %d", msgSize);
    NSDictionary* msg;
    NSString* sender; //sending device
    long serverTime;
    NSString* msgData;
    NSString* msgId;
    NSDateFormatter* format = [[NSDateFormatter alloc] init];
    [format setDateFormat:@"yyyy/MM/dd HH:mm:ss"];
    
    for(int i = 0; i < msgSize; i++) {
        msg = msgList[i];
        msgId = msg[@"msgId"];
        sender = msg[@"sender"];
        serverTime = [msg[@"serverTime"] longValue];
        NSDate* date = [NSDate dateWithTimeIntervalSince1970:(serverTime /1000)];
        msgData = msg[@"data"];
        NSString* displayStr = [NSString stringWithFormat:@"p2p:%@  [sent time:%@]", msgData, [format stringFromDate:date]];
        //[self.messages insertObject:displayStr atIndex:0];
        [self.messages addObject:displayStr];
    }
    
    [self.tView reloadData];
    NSIndexPath *topIndexPath = [NSIndexPath indexPathForRow:messages.count-1 inSection:0];
    [self.tView scrollToRowAtIndexPath:topIndexPath
                      atScrollPosition:UITableViewScrollPositionMiddle
                              animated:YES];
}


-(void) onUpstreamMsgResult: (int) resultCode resultMsg : (NSString*) resultMsg umi : (NSString*) umi {
    NSLog(@"onUpstreamMsgResult umi = %@", umi);
}


-(void) onP2PMsgResult: (int) resultCode resultMsg : (NSString*) resultMsg pmi : (NSString*) pmi {
    NSLog(@"onP2PMsgResult pmi = %@", pmi);
}



#pragma mark -
#pragma mark Table delegates

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSString *s = (NSString *) [messages objectAtIndex:indexPath.row];
    static NSString *CellIdentifier = @"ChatCellIdentifier";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    cell.textLabel.text = s;
    return cell;
}


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return messages.count;
}




@end

File: device-app/ios-swift/sample/MsgViewController.swift
//
//  MsgViewController.swift
//  sample
//
//  Created by 김정도 on 2018. 1. 26..
//  Copyright © 2018년 Rationalowl. All rights reserved.
//

import UIKit;



class MsgViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, MessageDelegate {
   
    
    @IBOutlet weak var msgListView: UIView!;
    @IBOutlet weak var inputMessageField: UITextField!
    @IBOutlet weak var tVew: UITableView!;
    
    var messages: Array<String>!;
    
    override func viewDidLoad() {
        super.viewDidLoad();
        
        messages = Array <String>();
        
        tVew.delegate = self;
        tVew.dataSource = self;
        
        // set message delegate
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.setMessageDelegate(self);
        
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning();
        // Dispose of any resources that can be recreated.
    }
    
    /////////////////////////////////////////////////////////////////
    // IBActions
    /////////////////////////////////////////////////////////////////
    
    @IBAction func sendUpstreamMsg() {
        let serverId: String = "e2c5f56d6cda41279c3a4ad8549052bf";
        let msg: String = inputMessageField.text!;
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.sendUpstreamMsg(msg, serverRegId:serverId);
        
        let displayStr: String = "=>\(msg)";
        self.messages.append(displayStr);
        self.tVew.reloadData();
        
        let topIndexPath = IndexPath(row: messages.count-1, section: 0)
        self.tVew.scrollToRow(at: topIndexPath, at: UITableView.ScrollPosition.middle, animated: true);
    }
    
    
    @IBAction func sendP2PstreamMsg() {
        let msg: String = inputMessageField.text!;
        let devices: Array<String> = ["a03e1768d6ab411b930eb6dd6b5048b4"];
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.sendP2PMsg(msg, devices:devices);
        
        let displayStr: String = "=>(p2p)\(msg)";
        self.messages.append(displayStr);
        self.tVew.reloadData();
        let topIndexPath = IndexPath(row: messages.count-1, section: 0)
        self.tVew.scrollToRow(at: topIndexPath, at: UITableView.ScrollPosition.middle, animated: true);
    }
    
    
    /////////////////////////////////////////////////////////////////
    // message delegate
    /////////////////////////////////////////////////////////////////
    
    func onDownstreamMsgRecieved(_ msgSize: Int32, msgList: [Any]!, alarmIdx: Int32) {
        print("onMsgRecieved msg size = \(msgSize)")
        var msg: Dictionary<String, Any>;
        // who send message
        var sender: String;
        var sendTime: Double;
        var msgData: String;
        let dateFormatter = DateFormatter();
        dateFormatter.dateFormat = "yyyy/MM/dd HH:mm:ss";
        
        for i in 0..<msgList.count {
            msg = msgList[i] as! Dictionary<String, Any>;
            // message sender(app server)'s app server registraion id
            sender = msg["sender"] as! String;
            // message sent time
            sendTime = msg["serverTime"] as! Double;
            let date = Date(timeIntervalSince1970: sendTime/1000);
            msgData = msg["data"] as! String;
            let displayStr: String = "\(msgData)   \(dateFormatter.string(from: date))";
            self.messages.append(displayStr);
        }
        
        self.tVew.reloadData();
        let topIndexPath = IndexPath(row: messages.count-1, section: 0)
        self.tVew.scrollToRow(at: topIndexPath, at: UITableView.ScrollPosition.middle, animated: true);
    }
    
    
    func onP2PMsgRecieved(_ msgSize: Int32, msgList: [Any]!, alarmIdx: Int32) {
        print("onMsgRecieved msg size = \(msgSize)")
        var msg: Dictionary<String, Any>;
        var sender: String;
        var sendTime: Double;
        var msgData: String;
        let dateFormatter = DateFormatter();
        dateFormatter.dateFormat = "yyyy/MM/dd HH:mm:ss";
        
        for i in 0..<msgList.count {
            msg = msgList[i] as! Dictionary<String, Any>;
            // message sender(device app)'s device app registraion id
            sender = msg["sender"] as! String;
            // message sent time
            sendTime = msg["serverTime"] as! Double;
            let date = Date(timeIntervalSince1970: sendTime/1000);
            msgData = msg["data"] as! String;
            let displayStr: String = "p2p:\(msgData)   \(dateFormatter.string(from: date))";
            self.messages.append(displayStr);
        }
        
        self.tVew.reloadData();
        let topIndexPath = IndexPath(row: messages.count-1, section: 0)
        self.tVew.scrollToRow(at: topIndexPath, at: UITableView.ScrollPosition.middle, animated: true);
    }
    
    
    func onUpstreamMsgResult(_ resultCode: Int32, resultMsg: String!, umi: String!) {
        print("onMsgRecieved msg upstream message id = \(umi)")
    }
    
    
    func onP2PMsgResult(_ resultCode: Int32, resultMsg: String!, pmi: String!) {
        print("onMsgRecieved msg p2p message id = \(pmi)")
    }
    
    
    /////////////////////////////////////////////////////////////////
    // UITableViewDataSource
    /////////////////////////////////////////////////////////////////
    
    func numberOfSectionsInTableView(in tableView: UITableView) -> Int {
        return 1
    }
    
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return messages.count;
    }
    
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let str: String = messages[indexPath.row];
        //let str: String = messages.object(at: indexPath.row) as! String;
        let cell: UITableViewCell = UITableViewCell(style: UITableViewCell.CellStyle.subtitle, reuseIdentifier: "Cell");
        //let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "ChatCellIdentifier")!;
        cell.textLabel?.text = str;
        return cell;
    }
}

